mlflow/tracking/client.py

```diff
@@ -1010,6 +1010,67 @@ def log_dict(self, run_id, dictionary, artifact_file):
                 else:
                     json.dump(dictionary, f, indent=2)
 
+    def log_figure(self, run_id, figure, artifact_file):
+        """
+        Log a figure as an artifact. The following figure objects are supported:
+
+        - `matplotlib.figure.Figure`_
+        - `plotly.graph_objects.Figure`_
+
+        .. _matplotlib.figure.Figure:
+            https://matplotlib.org/api/_as_gen/matplotlib.figure.Figure.html
+
+        .. _plotly.graph_objects.Figure:
+            https://plotly.com/python-api-reference/generated/plotly.graph_objects.Figure.html
+
+        :param run_id: String ID of the run.
+        :param figure: Figure to log.
+        :param artifact_file: The run-relative artifact file path in posixpath format to which
+                              the figure is saved (e.g. "dir/file.png").
+
+        .. code-block:: python
+            :caption: Matplotlib Example
+
+            import mlflow
+            import matplotlib.pyplot as plt
+
+            fig, ax = plt.subplots()
+            ax.plot([0, 1], [2, 3])
+
+            run = client.create_run(experiment_id="0")
+            client.log_figure(run.info.run_id, fig, "figure.png")
+
+        .. code-block:: python
+            :caption: Plotly Example
+
+            import mlflow
+            from plotly import graph_objects as go
+
+            fig = go.Figure(go.Scatter(x=[0, 1], y=[2, 3]))
+
+            run = client.create_run(experiment_id="0")
+            client.log_figure(run.info.run_id, fig, "figure.html")
+        """
+        MATPLOTLIB_FIGURE_CLASS_NAME = "matplotlib.figure.Figure"
+        PLOTLY_FIGURE_CLASS_NAME = "plotly.graph_objs._figure.Figure"
+
+        qual_class_name = "{}.{}".format(figure.__class__.__module__, figure.__class__.__name__)
+        if qual_class_name not in [MATPLOTLIB_FIGURE_CLASS_NAME, PLOTLY_FIGURE_CLASS_NAME]:
+            raise TypeError("Unsupported figure object type: '{}'".format(type(figure)))
+
+        with self._log_artifact_helper(run_id, artifact_file) as tmp_path:
+            # If we used `isinstance(figure, matplotlib.figure.Figure)` here, a user would need to
+            # install matplotlib to log a plotly figure. To avoid that, detect the object type
+            # with class name matching.
+            if qual_class_name == MATPLOTLIB_FIGURE_CLASS_NAME:
+                figure.savefig(tmp_path)
+            elif qual_class_name == PLOTLY_FIGURE_CLASS_NAME:
+                import plotly
+
+                plotly.offline.plot(
+                    figure, filename=tmp_path, include_plotlyjs="cdn", auto_open=False
+                )
```

Comment:
There might be a better approach...

Reply:
Can we write to a memory buffer like io.BytesIO instead of a local file, then use pass that buffer to the remote artifact logging request?

E.g. pass in the buffer to `figure.savefig` or `figure.write_html`

I realize you're probably asking whether there's a better way to inspect the type of figure and right way to write the plot rather than use of a temporary file... but we may be able to avoid use of temp files 🤷

Reply:
@pogil If `log_artifact` supported a memory buffer, we could do that but it seems `log_artifact` doesn't.

Reply:
Should we add that support for `log_artifact`?

Reply:
Offline discussion on 2020/11/23:

Updating `log_artifact` to support a memory buffer requires the following changes:

- The first argument of `log_artifact` is currently named as `local_path` which doesn't sound like it accepts a memory buffer. We'd need to rename it to something like `local_path_or_memory_buffer`.
- We'd need to add a new argument (e.g. `artiafact_file`) to specify the the artifact filename.